core/src/upgrade/mod.rs

```diff
@@ -145,7 +145,7 @@ pub trait InboundUpgrade<C>: UpgradeInfo {
     type Error;
     /// Future that performs the handshake with the remote.
     // TODO: remove Unpin
-    type Future: Future<Output = Result<Self::Output, Self::Error>> + Unpin;
+    type Future: Future<Output = Result<Self::Output, Self::Error>> + Unpin + Send;
```

Comment:
This neither.

Reply:
@tomaka I am sorry, I must be misunderstanding something here.

You would like `<NoiseAuthenticated<P, C, R> as InboundUpgrade<T>>::Future` to be `Send`, right?

As far as I understood something can only be `Send` if everything below it is `Send`.

`<NoiseAuthenticated<P, C, R> as InboundUpgrade<T>>::upgrade_inbound` calls `<NoiseConfig<P, C, R> as InboundUpgrade<T>>::upgrade_inbound` which by our requirement above needs to return a `Send` `Future`, thus the added trait bound (`Send`) on the `InboundUpgrade` trait.

What am I missing?

Reply:
As you mentioned, the `InboundUpgrade` is `Send` if all its components are `Send`. So all you have to do is require `Send` on its components within the code of `libp2p-noise`.

> As far as I understood something can only be Send if everything below it is Send.

The exception is `Box<dyn Future>` which is never send and that you have to replace with `Box<dyn Future + Send>`. That's the only change that you might have to do outside of libp2p-noise.